### PR TITLE
feat: Find BRS-021 receivers

### PR DESCRIPTION
**Pull request by MWO1024**
## Description

Find receivers of the forwarded metered data, based on the master data for the metering point. This also includes splitting the metered data so the receivers only receives metered data for the period they are applicable for.

## References

Link to assignment: https://app.zenhub.com/workspaces/mosaic-60a6105157304f00119be86e/issues/gh/energinet-datahub/team-mosaic/404

## Checklist

- [ ] Should the change be behind a feature flag?
- [ ] Can the feature be meaningfully disabled or circumvented if there are issues (e.g., database-breaking changes)?
- [ ] Has it been considered whether data is being delivered to the wrong actor?
- [ ] Subsystem test executed (dev_002/dev_003)
- [ ] Is there time to monitor state of the release to Production?
- [x] Reference to the task

### CODE REVIEW

File: source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/ElectricityMarket/MeteringPointReceiversProvider.cs [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        var currentMasterData = new MasterDataWithMeteredData(
+            MasterData: firstMasterData,
+            ValidFrom: firstMasterData.ValidFrom.ToInstant(),
+            ValidTo: firstMasterData.ValidTo.ToInstant(),
+            MeteredDataList: []);
+
+        List<MasterDataWithMeteredData> masterDataWithMeteredDataList = [currentMasterData];
+
+        foreach (var meteredData in sortedMeteredData.Values)
+        {
```
Comment (MWO1024) [inline source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/ElectricityMarket/MeteringPointReceiversProvider.cs:108]: I'm concerned about iterating over the points instead of iterating over the buckets. There can be around 105102 points per transaction and iterating over each of those seems inefficient compared to iterating over the buckets (we expect to have 1, maybe 2 or 3 of those)

File: source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/ElectricityMarket/MeteringPointReceiversProvider.cs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        }
+
+        if (!_resolutionToDurationMap.TryGetValue(currentResolution, out var durationForCurrentResolution))
+            throw new ArgumentOutOfRangeException(nameof(currentResolution), currentResolution.Name, "The resolution is not supported");
+
+        // Resolution should now be below a month, so we can use math directly on the instant (the resolution
+        // is a fixed amount of time)
+        var nextTimestamp = timestamp.Plus(durationForCurrentResolution);
+        return nextTimestamp;
+    }
```
Comment (MWO1024) [inline source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/ElectricityMarket/MeteringPointReceiversProvider.cs:172]: It would be nice to have a test of this, but ideally without having to go through all the other code. Just to make sure we haven't missed some silly edge case.
We can consider it at least. It just seems excessive (or maybe not?) to test we can split on multiple master data for each resolution.

File: source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/V1/Handlers/StartForwardMeteredDataHandlerV1.cs [hunk 3]
```diff
@@ -371,4 +378,19 @@ private Point MapPoints(ForwardMeteredDataInputV1.MeteredData eo)
...
+    //     DateTimeOffset StartDateTime,
+    //     DateTimeOffset EndDateTime,
+    //     IReadOnlyCollection<ReceiversWithMeteredDataV1.AcceptedMeteredData> MeteredData)
+    // {
+    //     public record AcceptedMeteredData(
+    //         int Position,
+    //         // TODO: Are these two optional?
+    //         decimal? EnergyQuantity,
+    //         Quality? QuantityQuality);
+    // }
```
Comment (MWO1024) [inline source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/V1/Handlers/StartForwardMeteredDataHandlerV1.cs:395]: We should be able to safely delete this now.

File: source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/ElectricityMarket/MeteringPointReceiversProvider.cs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+    private readonly DateTimeZone _dateTimeZone = dateTimeZone;
+
+    public List<ReceiversWithMeteredDataV1> GetReceiversWithMeteredDataFromMasterDataList(
+        IReadOnlyCollection<MeteringPointMasterData> meteringPointMasterDataList,
+        ForwardMeteredDataInputV1 input)
+    {
+        // Ensure metered data is sorted by position
+        var sortedMeteredData = new SortedDictionary<int, ReceiversWithMeteredDataV1.AcceptedMeteredData>(
+            input.MeteredDataList.ToDictionary(
```
[the remark this answers is not in the document]
Comment (ebbeknudsen) [reply source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/ElectricityMarket/MeteringPointReceiversProvider.cs:44]: The only way to create a `SortedList` or `SortedDictionary` is by creating a dictionary where the key is the property that is sorted by. I use a `SortedDictionary` to ensure that the received parameter is a sorted list, because if the data isn't sorted by position then the logic will fail.

File: source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/ElectricityMarket/MeteringPointReceiversProvider.cs (hunk 2 above)
Comment (ebbeknudsen) [reply source/ProcessManager.Orchestrations/Processes/BRS_021/ForwardMeteredData/ElectricityMarket/MeteringPointReceiversProvider.cs:172]: Good point, i'll make separate tests for each resolution.

**Review by MWO1024 — COMMENTED** [review]
LGTM

**Review by lasrinnil — APPROVED** [review]
